lib/elixir/lib/access.ex

```diff
@@ -422,4 +422,73 @@ defmodule Access do
   defp all([], _next, gets, updates) do
     {:lists.reverse(gets), :lists.reverse(updates)}
   end
+
+  @doc ~S"""
+  Accesses the element at `index` (zero based) of a list.
+
+  ## Examples
+
+      iex> list = [%{name: "john"}, %{name: "mary"}]
+      iex> get_in(list, [Access.at(1), :name])
+      "mary"
+      iex> get_and_update_in(list, [Access.at(0), :name], fn
+      ...>   prev -> {prev, String.upcase(prev)}
+      ...> end)
+      {"john", [%{name: "JOHN"}, %{name: "mary"}]}
+      iex> pop_in(list, [Access.at(0), :name])
+      {"john", [%{}, %{name: "mary"}]}
+
+  When the index is out of bounds, `nil` is returned and the update function is never called:
+      iex> list = [%{name: "john"}, %{name: "mary"}]
+      iex> get_in(list, [Access.at(10), :name])
+      nil
+      iex> get_and_update_in(list, [Access.at(10), :name], fn
+      ...>   prev -> {prev, String.upcase(prev)}
+      ...> end)
+      {nil, [%{name: "john"}, %{name: "mary"}]}
+
+  An error is raised for negative indexes:
+
+      iex> get_in([], [Access.at(-1)])
+      ** (RuntimeError) Access.at/1 expected a non-negative index, got: -1
+
+  An error is raised if the accessed structure is not a list:
+
+      iex> get_in(%{}, [Access.at(1)])
+      ** (RuntimeError) Access.at/1 expected a list, got: %{}
+  """
+  def at(index) when index >= 0 do
+    fn(op, data, next) -> at(op, data, index, next) end
+  end
+
+  def at(index) do
+    raise "Access.at/1 expected a non-negative index, got: #{inspect index}"
+  end
+
+  defp at(:get, data, index, next) when is_list(data) do
+    data |> Enum.at(index) |> next.()
+  end
+
+  defp at(:get_and_update, data, index, next) when is_list(data) do
+    at(:get_and_update, data, index, next, [])
+  end
+
+  defp at(_op, data, _index, _next) do
+    raise "Access.at/1 expected a list, got: #{inspect data}"
```

Comment:
I was following the same pattern used in `all`, `elem`… in the same file (no explicit ArgumentError, custom error messages).

I could leave it as-is or add the ArgumentError everywhere on the file so that it is consistent… what do you think?